src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -76,23 +87,129 @@ public void completed(Integer result, Void v) {
                     });
         } catch (IOException e) {
             VolleyLog.e(e, "Failed to read file %s", file.getAbsolutePath());
+            if (channel != null) {
+                try {
+                    channel.close();
+                } catch (IOException e1) {
+                    VolleyLog.e(e1, "Failed to close channel after IOException");
+                }
+            }
             callback.onGetComplete(null);
         }
     }
 
-    /** Returns a file object for the given cache key. */
-    File getFileForKey(String key) {
-        return new File(mRootDirectorySupplier.get(), DiskBasedCacheUtility.getFilenameForKey(key));
+    /** Puts the cache entry with a specified key into the cache. */
+    @Override
+    public void put(final String key, Cache.Entry entry, final OnPutCompleteCallback callback) {
+        if (DiskBasedCacheUtility.wouldBePruned(
+                mTotalSize, entry.data.length, mMaxCacheSizeInBytes)) {
+            return;
+        }
+
+        final File file = DiskBasedCacheUtility.getFileForKey(key, mRootDirectorySupplier);
+        Path path = Paths.get(file.getPath());
+
+        // channel we can close after IOException
+        AsynchronousFileChannel channel = null;
+        try {
+            final AsynchronousFileChannel afc =
+                    AsynchronousFileChannel.open(
+                            path, StandardOpenOption.WRITE, StandardOpenOption.CREATE);
+            channel = afc;
+            final CacheHeader header = new CacheHeader(key, entry);
+            int headerSize = header.getHeaderSize();
+            int size = entry.data.length + headerSize;
+            ByteBuffer buffer = ByteBuffer.allocate(size);
+            header.writeHeader(buffer);
+            buffer.put(entry.data);
+            buffer.flip();
+            afc.write(
+                    /* source= */ buffer,
+                    /* position= */ 0,
+                    /* attachment= */ null,
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer resultLen, Void ignore) {
+                            header.size = resultLen;
+                            mTotalSize =
+                                    DiskBasedCacheUtility.putEntry(
+                                            key, header, mTotalSize, mEntries);
+                            mTotalSize =
+                                    DiskBasedCacheUtility.pruneIfNeeded(
+                                            mTotalSize,
+                                            mMaxCacheSizeInBytes,
+                                            mEntries,
+                                            mRootDirectorySupplier);
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel after successful write");
```

Comment:
I don't think the write can be considered successful if the close fails - e.g. I don't think we should update total size or perform pruning until the close completes, and then only if it's successful.